File: TMG_USB_IO-Link_Interface_V2_DLL/Sample_x64/Sample_CSharp/frmMain.cs
using System;
using System.Diagnostics;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using TMGIOLUSBIF20_64_DOTNET;
using System.IO;

namespace USBIOL_DLL_Sample_CSharp
{
  public partial class frmMain : Form
  {
    public frmMain()
    {
      InitializeComponent();
    }

    clsTMGIOLUSBIF20 IF20 = new clsTMGIOLUSBIF20();
    clsTMGIOLUSBIF20.clsUSBIOLMaster IOLM = null;
    int ActivePort = 0;
    clsTMGIOLUSBIF20.clsFwUpdateStatus updatestatus = new clsTMGIOLUSBIF20.clsFwUpdateStatus();

    private void frmMain_Load(object sender, EventArgs e)
    {
      clsTMGIOLUSBIF20.eError result = clsTMGIOLUSBIF20.eError.OK;
      tbMessages.Text = "";

      //Read DLL Info
      clsTMGIOLUSBIF20.clsDllInfo DLLInfo = new clsTMGIOLUSBIF20.clsDllInfo();

      WriteInfo("Info about IOLUSBIF20.DLL:");
      WriteInfo("    Revision of C DLL:       " + DLLInfo.C_DLL_Version + "_" + DLLInfo.C_DLL_Build);
      WriteInfo("    Revision of .NET DLL:    " + DLLInfo.NET_DLL_Version);
      WriteInfo();

      //search on USB Masters
      clsTMGIOLUSBIF20.clsUSBIOLMaster[] IOLMList = IF20.GetUSBMaster();
      if (IOLMList.Length == 0)
      {
        result = clsTMGIOLUSBIF20.eError.DEVICE_NOT_AVAILABLE;
        goto RError;
      }
      //you could implement a combobox hier to choose the master. In the sample the first is taken.
      IOLM = IOLMList[0];
      if (IOLM.IOL_Create() == clsTMGIOLUSBIF20.eError.OK)
      {
        //Read Master Info
        WriteInfo("IO-Link USB V2 Master");
        WriteInfo("    Firmware Revision:      " + IOLM.RevisionFirmware);
        WriteInfo("    IO-Link Stack Revision: " + IOLM.RevisionIOLStack);
      }
      return;
RError:
      if (result == clsTMGIOLUSBIF20.eError.DEVICE_NOT_AVAILABLE)
      {
        MessageBox.Show("Error Message: " + result.ToString() + "\r\n" +
           "Maybe no driver is installed" + "\r\n" +
           "Please install the driver in menu <Options/Install Drivers>" + "\r\n" +
           "and start the program again", "Critical Warning");
      }
      else
      {
        MessageBox.Show("Error Message: " + result.ToString(), "Critical Warning");
        Environment.Exit(0);
      }
    }

    private void WriteInfo(string text)
    {
      tbInfo.AppendText(text + "\r\n");
      tbInfo.ScrollToCaret();
    }

    private void WriteInfo()
    {
      tbInfo.AppendText("\r\n");
      tbInfo.ScrollToCaret();
    }

    private void Write(string text)
    {
      tbMessages.AppendText(text + "\r\n");
      tbMessages.ScrollToCaret();
    }

    private void Write()
    {
      tbMessages.AppendText("\r\n");
      tbMessages.ScrollToCaret();
    }

    private void timer1_Tick(object sender, EventArgs e)
    {
      timer1.Enabled = false;
      clsTMGIOLUSBIF20.eError result = IOLM.Port[ActivePort].IOL_ReadInputs();
      if (result == clsTMGIOLUSBIF20.eError.INTERNAL_ERROR)
      {  
        MessageBox.Show("Connection lost to the connected USB IO-Link Master", "Critical Warning");
        Environment.Exit(0);
      }
      tbPDIn.Text = IF20.Utils.GetHexStringFromBytes(ref IOLM.Port[ActivePort].Inputs.Data);
      if (IOLM.Port[ActivePort].DeviceState.EventAvailable)
      {
        IOLM.LastEvent.IOL_ReadEvent(ref IOLM.Port[ActivePort].DeviceState);
        Write("Event: " + IOLM.LastEvent.Instance + " " + 
                      IOLM.LastEvent.Type + " " + 
                      IOLM.LastEvent.Mode + " " +
                      IOLM.LastEvent.EventCode);
      }
      result = IOLM.Port[ActivePort].IOL_ReadOutputs();
      tbPDOutRead.Text = IF20.Utils.GetHexStringFromBytes(ref IOLM.Port[ActivePort].Outputs.Data);
      timer1.Enabled = true;
    }

    private void cmdRead_Click_1(object sender, EventArgs e)
    {
      byte[] data = null;
      int ErrorCode = 0;
      IOLM.Port[ActivePort].IOL_ReadReq(Convert.ToInt32(tbIndexRead.Text), Convert.ToInt32(tbSubindexRead.Text), ref data, ref ErrorCode);
      if (ErrorCode == 0)
      {
        tbDataRead.Text = IF20.Utils.GetHexStringFromBytes(ref data);
      }
      else
      {
        tbDataRead.Text = "";
      }
      tbReadError.Text = ErrorCode.ToString();
    }

    private void cmdWrite_Click(object sender, EventArgs e)
    {
      byte[] data = IF20.Utils.GetBytesFromTextHex(tbDataWrite.Text);
      int ErrorCode = 0;
      clsTMGIOLUSBIF20.eError result = IOLM.Port[ActivePort].IOL_WriteReq(Convert.ToInt32(tbIndexWrite.Text), Convert.ToInt32(tbSubindexWrite.Text), ref data, ref ErrorCode);
      tbDataRead.Text = "";
      tbReadError.Text = "";
      tbWriteError.Text = ErrorCode.ToString();
    }

    private void tbPDOutWrite_KeyPress(object sender, KeyPressEventArgs e)
    {
      if (e.KeyChar == '\r')
      {
        byte[] Outputs = IF20.Utils.GetBytesFromTextHex(tbPDOutWrite.Text);
        Outputs.CopyTo(IOLM.Port[ActivePort].Outputs.Data, Outputs.Length);
        clsTMGIOLUSBIF20.eError result = IOLM.Port[ActivePort].IOL_WriteOutputs();
        tbPDOutWrite.Text = IF20.Utils.GetHexStringFromBytes(ref Outputs);
      }
    }

    private void cmdPort1_Click(object sender, EventArgs e)
    {
      GroupBox1.Text = "IO-Link Port 1 - Pin 4";
      tbMessages.Text = "";
      ActivatePort(0);
    }

    private void cmdPort1_Deactivate_Click(object sender, EventArgs e)
    {
      IOLM.Port[0].Configuration.Deactivate();
      tbMessages.Text = "";
    }

    private void cmdPort1Validate_Click(object sender, EventArgs e)
    {
      IOLM.Port[0].Configuration.ConfiguredMode = clsTMGIOLUSBIF20.eConfiguredMode.IOLINK_OPERATE;
      IOLM.Port[0].Configuration.InspectionLevel = clsTMGIOLUSBIF20.eInspectionLevel.COMPATIBLE;
      IOLM.Port[0].Configuration.VendorID = 815;
      IOLM.Port[0].Configuration.DeviceID = 4711;
      clsTMGIOLUSBIF20.eError result = IOLM.Port[0].Configuration.IOL_SetPortConfig();
      if (result == clsTMGIOLUSBIF20.eError.OK)
      {
        //Wait some time for master detecting the device

        System.Threading.Thread.Sleep(1000);

        GetDeviceInfo(ref IOLM, ActivePort);

        timer1.Enabled = true;
      }
      else
      {
        MessageBox.Show("Error Message: " + result.ToString(), "Critical Warning");
      }
    }

    private void cmdPort2_Click(object sender, EventArgs e)
    {
      GroupBox1.Text = "IO-Link Port 2 - Pin 2";
      tbMessages.Text = "";
      ActivatePort(1);
    }

    private void cmdPort2_Deactivate_Click(object sender, EventArgs e)
    {
      IOLM.Port[1].Configuration.Deactivate();
      tbMessages.Text = "";
    }

    private void cmdRead_Click(object sender, EventArgs e)
    {
      byte[] data = null;
      int ErrorCode = 0;
      IOLM.Port[ActivePort].IOL_ReadReq(Convert.ToInt32(tbIndexRead.Text), Convert.ToInt32(tbSubindexRead.Text), ref data, ref ErrorCode);
      if (ErrorCode == 0)
        tbDataRead.Text = IF20.Utils.GetHexStringFromBytes(ref data);
      else
        tbDataRead.Text = "";
  
      tbReadError.Text = ErrorCode.ToString();
    }

    private void cmdWrite_Click_1(object sender, EventArgs e)
    {
      byte[] data = IF20.Utils.GetBytesFromTextHex(tbDataWrite.Text);
      int ErrorCode = 0;
      clsTMGIOLUSBIF20.eError result = IOLM.Port[ActivePort].IOL_WriteReq(Convert.ToInt32(tbIndexWrite.Text), Convert.ToInt32(tbSubindexWrite.Text), ref data, ref ErrorCode);
      tbDataRead.Text = "";
      tbReadError.Text = "";
      tbWriteError.Text = ErrorCode.ToString();
    }

    private void ActivatePort(int pActivePort)
    {
      ActivePort = pActivePort;
      IOLM.Port[ActivePort].Configuration.Clear();
      IOLM.Port[ActivePort].Configuration.ConfiguredMode = clsTMGIOLUSBIF20.eConfiguredMode.IOLINK_OPERATE;
      clsTMGIOLUSBIF20.eError result = IOLM.Port[ActivePort].Configuration.IOL_SetPortConfig();

      if (result != clsTMGIOLUSBIF20.eError.OK)
        goto ActivateError;

      //Wait some time for master detecting the device

      System.Threading.Thread.Sleep(1000);

      GetDeviceInfo(ref IOLM, ActivePort);

      timer1.Enabled = true;

      return;
ActivateError:
      MessageBox.Show("Error Message: " + result, "Critical Warning");
    }

    private void GetDeviceInfo(ref clsTMGIOLUSBIF20.clsUSBIOLMaster Master, int Port)
    {
      //Read Info from Device with Directparameterpage
      IOLM.Port[Port].Mode.IOL_GetMode(false);


      if (IOLM.Port[Port].DeviceState.DeviceState != clsTMGIOLUSBIF20.eDeviceState.LOST)
      {
        Write("IO-Link Device connected:");
        Write("   VendorId:                  " + IOLM.Port[Port].Mode.DirectParameterPage1.VendorId());
        Write("   DeviceId:                  " + IOLM.Port[Port].Mode.DirectParameterPage1.DeviceId());
        Write("   IO-Link Revision:          " + IOLM.Port[Port].Mode.DirectParameterPage1.RevisionId());
        Write("   ProcessData Input Length:  " + IOLM.Port[Port].Mode.DirectParameterPage1.PDInLength());
        Write("   ProcessData Output Length: " + IOLM.Port[Port].Mode.DirectParameterPage1.PDOutLength());
        Write("   SIO mode supported:        " + IOLM.Port[Port].Mode.DirectParameterPage1.SIOModeSupported());
      }
      else
      {
        Write("No Device Connected");
      }
      Write();
    }
    
    private void TSM_About_Click(object sender, EventArgs e)
    {
      MessageBox.Show("Copyright (c) 2014 TMG Technologie und Engineering GmbH" + "\r\n" +
         "Sample Application for IOLUSBIF20_DOTNET.DLL" + "\r\n" +
         "www.tmgte.de", "information");
    }

    private void frmMain_Closing(object sender, FormClosingEventArgs e)
    {
      if (IOLM != null)
      {
        IOLM.Port[0].Configuration.Deactivate();
        IOLM.Port[1].Configuration.Deactivate();
        IOLM.IOL_Destroy();
      }
    }
    private string GetOpenFilename()
    {
      OpenFileDialog fd  = new OpenFileDialog();
      fd.Filter = "firmware update files (*.xml)|*.xml|All files (*.*)|*.*";
      fd.FilterIndex = 2;
      fd.RestoreDirectory = true;
      if (fd.ShowDialog() == DialogResult.OK)
        return fd.FileName;
      else
        return null;
    }

    private void cmdUpdate_Click(object sender, EventArgs e)
    {
      if (IOLM != null)
      {
        string firmwareFile = GetOpenFilename();
        if (firmwareFile != "")
        {
          clsTMGIOLUSBIF20.clsFwUpdateInfo updateinfo = new clsTMGIOLUSBIF20.clsFwUpdateInfo();
          clsTMGIOLUSBIF20.eFwUpdateReturnValue retval = IOLM.Port[0].FirmwareUpdate.StartByMetafile(firmwareFile, ref updateinfo, ref updatestatus);
          if (updatestatus.nextState != clsTMGIOLUSBIF20.eFwUpdateState.FWUPDATE_STATE_IDLE)
          {
            //Update wurde gestartet.
            tbMessages.Text = "";
            //update infos anzeigen
            Write("Starting Firmware Update:");
            Write("Meta file name=" + firmwareFile);
            Write("Firmware Length=" + updateinfo.fwLength);
            Write("used hardware key=:" + updateinfo.hwKey);
            Write("vendor ID=" + updateinfo.vendorID);
            Write("password required:" + updateinfo.fwPasswordRequired.ToString() + "\n\r");
            //update status anzeigen
            printUpdateStatus(retval, updatestatus);
            //timer starten
            timer1.Enabled = false; // wenn anderer Timer läuft, diesen stoppen
            timerUpdate.Enabled = true; // timer starten
          }
          else
          {
            //update status und fehler anzeigen
            //update status anzeigen
            tbMessages.Text = "";
            Write("Starting Firmware Update:" + "\n\r");
            Write("Meta file name=" + firmwareFile + "\n\r");
            printUpdateStatus(retval, updatestatus);
          }
        }
      }
    }

    private void printUpdateStatus(clsTMGIOLUSBIF20.eFwUpdateReturnValue retval  , clsTMGIOLUSBIF20.clsFwUpdateStatus   updatestatus )
    {
      string text="";
      if ((retval != clsTMGIOLUSBIF20.eFwUpdateReturnValue.FWUPDATE_RET_OK) || (updatestatus.executedState != updatestatus.nextState))
        text = "FWUP: From=" + updatestatus.executedState.ToString() + " Next=" + updatestatus.nextState.ToString() + " retval=" + retval.ToString() + " dllretval=" + updatestatus.dllReturnValue.ToString() + "\n\r";
      else
        if ((updatestatus.BlobStatus.executedState != clsTMGIOLUSBIF20.eBlobState.BLOB_STATE_IDLE) || (updatestatus.BlobStatus.nextState != clsTMGIOLUSBIF20.eBlobState.BLOB_STATE_IDLE))
          text += "BLOB: From=" + updatestatus.BlobStatus.executedState.ToString() + " Next=" + updatestatus.BlobStatus.nextState + " blobretval=" + updatestatus.blobReturnValue.ToString() + " PercenComplete=" + updatestatus.BlobStatus.PercentComplete + "\n\r";
      if (text != "")
      {
        Write(text);
      }
    }

    private void timerUpdate_Tick_1(object sender, EventArgs e)
    {
      //zuerst timer stoppen, damit sich nichts überlappt
      timerUpdate.Enabled = false;
      string password = "";
      if (updatestatus.nextState == clsTMGIOLUSBIF20.eFwUpdateState.FWUPDATE_STATE_PASSWORD)
      {
        password = Interaction.InputBox("enter Passwort", "Firmware Update", "", -1, -1);
      }
      clsTMGIOLUSBIF20.eFwUpdateReturnValue retval = IOLM.Port[0].FirmwareUpdate.IOL_FwUpdateContinue(password, ref updatestatus);
      printUpdateStatus(retval, updatestatus);
      //am schluss timer wieder starten
      if (updatestatus.nextState != clsTMGIOLUSBIF20.eFwUpdateState.FWUPDATE_STATE_IDLE)
        timerUpdate.Enabled = true;
    }

  private void sendpatterns()
  {
    System.IO.Ports.SerialPort ComPort = new System.IO.Ports.SerialPort();
    ComPort.PortName = "COM141";
    ComPort.BaudRate = 38400;
    ComPort.DataBits = 8;
    ComPort.Parity = System.IO.Ports.Parity.None;
    ComPort.ParityReplace = 63;
    ComPort.StopBits = System.IO.Ports.StopBits.One;
    ComPort.Handshake = System.IO.Ports.Handshake.None;
    ComPort.ReadBufferSize = 128000;
    ComPort.ReadTimeout = -1;
    ComPort.ReceivedBytesThreshold = 1;
    ComPort.RtsEnable = false;
    ComPort.WriteBufferSize = 200;
    ComPort.WriteTimeout = -1;
    ComPort.DiscardNull = false;
    ComPort.DtrEnable = false;
    ComPort.Open();
    System.Threading.Thread.Sleep(1000);
    byte[] x = { 0, 1, 2, 3 };
    ComPort.Write(x, 0, 4);
    System.Threading.Thread.Sleep(1000);
    char[] y = { 'O', 'K', '\0' };
    ComPort.Write(y, 0, 3);

    ComPort.Close();
  }
 
  }
}

File: TMG_USB_IO-Link_Interface_V2_DLL/Sample_x86/Sample_VB/Source/frmMain.vb
Imports TMGIOLUSBIF20_DOTNET
Imports TMGIOLUSBIF20_DOTNET.clsTMGIOLUSBIF20

Public Class frmMain

  Dim IF20 As New clsTMGIOLUSBIF20
  Dim IOLM As clsUSBIOLMaster = Nothing
  Dim Timer1Count As Integer = 0
  Dim ActivePort As Integer = 0
  Dim updatestatus As New clsFwUpdateStatus

  Private Sub frmMain_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

    Dim result As eError = eError.OK
    tbMessages.Text = ""

    'DLL Info abrufen
    Dim DLLInfo As New clsDllInfo

    WriteInfo("Info about the IOLUSBIF20.DLL:")
    WriteInfo("    Revision of C DLL:       " & DLLInfo.C_DLL_Version & "_" & DLLInfo.C_DLL_Build)
    WriteInfo("    Revision of .NET DLL:    " & DLLInfo.NET_DLL_Version)
    WriteInfo()

    'search for USB Masters 
    Dim IOLMList() As clsUSBIOLMaster = IF20.GetUSBMaster()
    If IOLMList.Length = 0 Then result = eError.DEVICE_NOT_AVAILABLE : GoTo RError
    'you could implement a combobox hier to choose the master. In the sample the first is taken.
    IOLM = IOLMList(0)
    If IOLM.IOL_Create() = eError.OK Then
      'Read Master Info
      WriteInfo("IO-Link USB V2 Master")
      WriteInfo("    Firmware Revision:      " & IOLM.RevisionFirmware)
      WriteInfo("    IO-Link Stack Revision: " & IOLM.RevisionIOLStack)
    End If
    Return
RError:
    If result = eError.DEVICE_NOT_AVAILABLE Then
      MsgBox("Error Message: " & result.ToString & vbCrLf &
             "Maybe no driver is installed" & vbCrLf &
             "Please install the driver in menu <Options/Install Drivers>" & vbCrLf &
             "and start the program again", MsgBoxStyle.Critical)
    Else
      MsgBox("Error Message: " & result.ToString, MsgBoxStyle.Critical)
      End
    End If
  End Sub

  Private Sub frmMain_FormClosed(ByVal sender As Object, ByVal e As System.Windows.Forms.FormClosedEventArgs) Handles Me.FormClosed
    If Not IsNothing(IOLM) Then
      IOLM.Port(0).Configuration.Deactivate()
      IOLM.Port(1).Configuration.Deactivate()
      IOLM.IOL_Destroy()
    End If
  End Sub

  Private Sub cmdPort1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdPort1.Click
    GroupBox1.Text = "IO-Link Port 1 - Pin 4"
    tbMessages.Text = ""
    ActivatePort(0)
  End Sub

  Private Sub cmdPort1_Deactivate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdPort1_Deactivate.Click
    IOLM.Port(0).Configuration.Deactivate()
    tbMessages.Text = ""
  End Sub

  Private Sub cmdPort1Validate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdPort1Validate.Click
    IOLM.Port(0).Configuration.ConfiguredMode = eConfiguredMode.IOLINK_OPERATE
    IOLM.Port(0).Configuration.InspectionLevel = eInspectionLevel.COMPATIBLE
    IOLM.Port(0).Configuration.VendorID = 335
    IOLM.Port(0).Configuration.DeviceID = 16
    IOLM.Port(0).Configuration.DSConfigure = eDSConfigure.ENABLED
    Dim result As eError = IOLM.Port(0).Configuration.IOL_SetPortConfig()
    If result = eError.OK Then

      'Wait some time for master detecting the device
      System.Threading.Thread.Sleep(1000)

      GetDeviceInfo(IOLM, ActivePort)

      Timer1.Enabled = True
    Else
      MsgBox("Error Message: " & result.ToString, MsgBoxStyle.Critical)
    End If

  End Sub

  Private Sub cmdPort2_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdPort2.Click
    GroupBox1.Text = "IO-Link Port 2 - Pin 2"
    tbMessages.Text = ""
    ActivatePort(1)
  End Sub

  Private Sub cmdPort2_Deactivate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdPort2_Deactivate.Click
    IOLM.Port(1).Configuration.Deactivate()
    tbMessages.Text = ""
  End Sub

  Private Sub ActivatePort(ByVal pActivePort As Integer)
    ActivePort = pActivePort
    IOLM.Port(ActivePort).Configuration.Clear()
    IOLM.Port(ActivePort).Configuration.ConfiguredMode = eConfiguredMode.IOLINK_OPERATE
    Dim result As eError = IOLM.Port(ActivePort).Configuration.IOL_SetPortConfig

    If result <> eError.OK Then GoTo ActivateError

    'Wait some time for master detecting the device
    System.Threading.Thread.Sleep(1000)

    GetDeviceInfo(IOLM, ActivePort)

    Timer1.Enabled = True

    Return
ActivateError:
    MsgBox("Error Message: " & result.ToString, MsgBoxStyle.Critical)
  End Sub

  Private Sub GetDeviceInfo(ByRef Master As clsTMGIOLUSBIF20.clsUSBIOLMaster, ByVal Port As Integer)

    'Informationen vom Device lesen mit Directparameterpage
    IOLM.Port(Port).Mode.IOL_GetMode(False)
    With IOLM.Port(Port).Mode
      If .DeviceState.DeviceState <> eDeviceState.LOST Then
        Write("IO-Link Device connected:")
        Write("   VendorId:                        " & .DirectParameterPage1.VendorId)
        Write("   DeviceId:                        " & .DirectParameterPage1.DeviceId)
        Write("   IO-Link Revision:                " & .DirectParameterPage1.RevisionId.ToString)
        Write("   ProcessData Input Length (bit):  " & .DirectParameterPage1.PDInLength)
        Write("   ProcessData Output Length (bit): " & .DirectParameterPage1.PDOutLength)
        Write("   SIO mode supported:              " & .DirectParameterPage1.SIOModeSupported)
      Else
        Write("No device connected")
      End If
      Write()
    End With
  End Sub

  Private Sub Timer1_Tick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Timer1.Tick
    Timer1.Enabled = False
    Dim result As eError = IOLM.IOL_ReadAllInputs()
    If result = eError.INTERNAL_ERROR Then
      MsgBox("Connection lost to the connected USB IO-Link Master", MsgBoxStyle.Critical)
      End
    End If
    tbPDIn.Text = IF20.Utils.GetHexStringFromBytes(IOLM.Port(ActivePort).Inputs.Data)
    If IOLM.Port(ActivePort).DeviceState.EventAvailable Then
      IOLM.LastEvent.IOL_ReadEvent(IOLM.Port(ActivePort).DeviceState)
      Write("Event: " & IOLM.LastEvent.Instance.ToString & " " &
                        IOLM.LastEvent.Type.ToString & " " &
                        IOLM.LastEvent.Mode.ToString & " " &
                        IOLM.LastEvent.EventCode.ToString)
    End If
    result = IOLM.Port(ActivePort).IOL_ReadOutputs()
    tbPDOutRead.Text = IF20.Utils.GetHexStringFromBytes(IOLM.Port(ActivePort).Outputs.Data, IOLM.Port(ActivePort).Outputs.Length)
    Dim info As New TMGIOLUSBIF20_DOTNET.clsTMGIOLUSBIF20.clsUSBIOLMaster.CHardwareInfo
    result = IOLM.IOL_GetHWInfo(info)
    If result = eError.OK Then
      Me.CheckBox1.Checked = info.isExternalPower
      Me.TextBox1.Text = info.PowerLevel.ToString
    End If
    Timer1.Enabled = True
  End Sub

  Private Sub tbPDOutWrite_KeyPress(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles tbPDOutWrite.KeyPress
    If e.KeyChar = Chr(13) Then
      Dim Outputs() As Byte = IF20.Utils.GetBytesFromTextHex(tbPDOutWrite.Text)
      For i As Integer = 0 To Outputs.Length - 1
        IOLM.Port(ActivePort).Outputs.Data(i) = Outputs(i)
      Next
      Dim result As eError = IOLM.Port(ActivePort).IOL_WriteOutputs()
      tbPDOutWrite.Text = IF20.Utils.GetHexStringFromBytes(Outputs)
    End If
  End Sub

  Private Sub cmdRead_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdRead.Click
    Dim data() As Byte = Nothing
    Dim ErrorCode As Integer
    Dim result As eError = IOLM.Port(ActivePort).IOL_ReadReq(CInt(tbIndexRead.Text), CInt(tbSubindexRead.Text), data, ErrorCode)
    If result = 0 Then
      tbDataRead.Text = IF20.Utils.GetHexStringFromBytes(data)
    Else
      tbDataRead.Text = "error:" & Hex(ErrorCode)
    End If
    tbReadError.Text = ErrorCode.ToString
  End Sub

  Private Sub cmdWrite_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdWrite.Click
    Dim data() As Byte = IF20.Utils.GetBytesFromTextHex(tbDataWrite.Text)
    Dim ErrorCode As Integer
    Dim result As eError = IOLM.Port(ActivePort).IOL_WriteReq(CInt(tbIndexWrite.Text), CInt(tbSubindexWrite.Text), data, ErrorCode)
    tbDataRead.Text = ""
    tbReadError.Text = ""
    tbWriteError.Text = ErrorCode.ToString
  End Sub

  Private Sub WriteInfo(ByVal text As String)
    tbInfo.AppendText(text & vbCrLf)
    tbInfo.ScrollToCaret()
  End Sub
  Private Sub WriteInfo()
    tbInfo.AppendText(vbCrLf)
    tbInfo.ScrollToCaret()
  End Sub
  Private Sub Write(ByVal text As String)
    tbMessages.AppendText(text & vbCrLf)
    tbMessages.ScrollToCaret()
  End Sub
  Private Sub Write()
    tbMessages.AppendText(vbCrLf)
    tbMessages.ScrollToCaret()
  End Sub


  Private Sub TSM_About_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles TSM_About.Click
    MsgBox("Copyright (c) 2013-2023 TMG Technologie und Engineering GmbH" & vbCrLf &
           "Sample Application for IOLUSBIF20_DOTNET.DLL" & vbCrLf &
           "www.tmgte.de", MsgBoxStyle.Information)
  End Sub

  Private Function GetOpenFilename() As String
    Dim fd As OpenFileDialog = New OpenFileDialog()
    fd.Filter = "firmware update files (*.xml)|*.xml|All files (*.*)|*.*"
    fd.FilterIndex = 2
    fd.RestoreDirectory = True
    If fd.ShowDialog() = Windows.Forms.DialogResult.OK Then
      Return fd.FileName
    Else
      Return Nothing
    End If
  End Function
  Private Sub CmdFirmwareUpdate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles CmdFirmwareUpdate.Click
    'Dim firmwareFile As String = FileSystem.
    Dim firmwareFile As String = GetOpenFilename()
    If IsNothing(firmwareFile) OrElse firmwareFile = "" Then Return
    Dim updateinfo As New clsFwUpdateInfo
    Dim retval As eFwUpdateReturnValue = IOLM.Port(0).FirmwareUpdate.StartByMetafile(firmwareFile, updateinfo, updatestatus)
    If updatestatus.nextState <> eFwUpdateState.FWUPDATE_STATE_IDLE Then
      'Update wurde gestartet.
      tbMessages.Text = ""
      'update infos anzeigen
      tbMessages.AppendText("Starting Firmware Update:" & vbCrLf)
      tbMessages.AppendText("Meta file name=" & firmwareFile & vbCrLf)
      tbMessages.AppendText("Firmware Length=" & updateinfo.fwLength & vbCrLf)
      tbMessages.AppendText("used hardware key=:" & updateinfo.hwKey & vbCrLf)
      tbMessages.AppendText("vendor ID=" & updateinfo.vendorID & vbCrLf)
      tbMessages.AppendText("password required:" & updateinfo.fwPasswordRequired.ToString & vbCrLf)
      'update status anzeigen
      printUpdateStatus(retval, updatestatus)
      'timer starten
      Timer1.Enabled = False ' wenn anderer Timer läuft, diesen stoppen
      TimerUpdate.Enabled = True ' timer starten
    Else
      'update status und fehler anzeigen
      'update status anzeigen
      tbMessages.Text = ""
      tbMessages.AppendText("Starting Firmware Update:" & vbCrLf)
      tbMessages.AppendText("Meta file name=" & firmwareFile & vbCrLf)
      printUpdateStatus(retval, updatestatus)
    End If

  End Sub

  Private Sub printUpdateStatus(ByVal retval As eFwUpdateReturnValue, ByRef updatestatus As clsFwUpdateStatus)
    Dim text As String = ""
    If (retval <> eFwUpdateReturnValue.FWUPDATE_RET_OK) Or (updatestatus.executedState <> updatestatus.nextState) Then
      text = "FWUP: From=" & updatestatus.executedState.ToString & " Next=" & updatestatus.nextState.ToString & " retval=" & retval.ToString & " dllretval=" & updatestatus.dllReturnValue.ToString & vbCrLf
    End If
    If (updatestatus.BlobStatus.executedState <> eBlobState.BLOB_STATE_IDLE) Or (updatestatus.BlobStatus.nextState <> eBlobState.BLOB_STATE_IDLE) Then
      text &= "BLOB: From=" & updatestatus.BlobStatus.executedState.ToString & " Next=" & updatestatus.BlobStatus.nextState & " blobretval=" & updatestatus.blobReturnValue.ToString & " PercenComplete=" & updatestatus.BlobStatus.PercentComplete & vbCrLf
    End If
    If (text <> "") Then
      tbMessages.AppendText(text & vbCrLf)
      tbMessages.ScrollToCaret()
    End If
  End Sub

  Private Sub TimerUpdate_Tick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles TimerUpdate.Tick
    'zuerst timer stoppen, damit sich nichts überlappt
    TimerUpdate.Enabled = False
    Dim password As String = ""
    If updatestatus.nextState = eFwUpdateState.FWUPDATE_STATE_PASSWORD Then
      password = InputBox("enter Passwort", "Firmware Update", "")
    End If
    Dim retval As eFwUpdateReturnValue = IOLM.Port(0).FirmwareUpdate.IOL_FwUpdateContinue(password, updatestatus)
    printUpdateStatus(retval, updatestatus)
    'am schluss timer wieder starten
    If updatestatus.nextState <> eFwUpdateState.FWUPDATE_STATE_IDLE Then TimerUpdate.Enabled = True
  End Sub

End Class
